net/instaweb/rewriter/csp.cc

```diff
@@ -172,12 +175,158 @@ CspSourceExpression CspSourceExpression::Parse(StringPiece input) {
       return CspSourceExpression();
     }
 
-    input.CopyToString(&result.mutable_url_data()->path_part);
+    // Normalize and tokenize the path.
+    StringPieceVector components;
+    SplitStringPieceToVector(input, "/", &components, true);
+
+    for (StringPiece c : components) {
+      GoogleString canon = GoogleUrl::CanonicalizePath(c);
+      if (canon.empty()) {
+        LOG(DFATAL) << "Path canonicalization returned empty string?" << c;
+        return CspSourceExpression();
+      }
+      result.mutable_url_data()->path_part.push_back(canon.substr(1));
+    }
+    result.mutable_url_data()->path_exact_match =
+        !input.empty() && !input.ends_with("/");
   }
 
   return result;
 }
 
+bool CspSourceExpression::Matches(
+    const GoogleUrl& origin_url, const GoogleUrl& url) const {
+  // Implementation of the "Does url match expression in origin with
+  // redirect count?" algorithm (where redirect count is 0 for our
+  // purposes, since we check the request).
+  // TODO(morlovich): This means redirect following may require changes
+  // here.
+  if (kind_ != kSelf && kind_ != kSchemeSource && kind_ != kHostSource) {
+    return false;
+  }
+
+  if (!origin_url.IsAnyValid() || !url.IsAnyValid()) {
```

Comment:
Hmm. Actually I want to limit this to http[s], since I dropped support for ws/wss (as we don't use them and they were making for some complicated expressions)